classes/log_fetcher/course_log_fetcher.php

```diff
@@ -0,0 +1,235 @@
+                    fputcsv($courseloghandles[$entrycourseid], $row);
+
+                    echo "Found log entry for course $entrycourseid \n";
+
+                    if ($row[1] == '\core\event\course_deleted') {
+                        $completedlogs++;
+
+                        // We've parsed all logs needed.
+                        if ($completedlogs == count($courseids)) {
+                            echo "Finishing early, {$completedlogs} course delete events found \n";
+                            fclose($loghandle);
+                            return;
+                        }
+                    }
+                }
+            }
+
+            fclose($loghandle);
+        }
+    }
+
+    /**
+     * Determines if log row is a course log entry we want.
+     * Matches courseid = contextinstanceid and contextlevel = course
+     *
+     * @param array $courseids Course ids.
+     * @param array $row       Log row.
+     * @return boolean
+     */
+    private static function is_course_log_entry($courseids, $row) {
+        if (in_array($row[11], $courseids) && $row[10] == CONTEXT_COURSE) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Downloads an s3 log file to the temp dir and opens a handle for it.
+     *
+     * @param string $s3logkey
+     * @return resource
+     */
+    private static function download_s3_log_file($s3logkey) {
+        $s3client = new s3_client();
+        $tempdir = make_temp_directory('s3logs_download');
+        $tempfilepath = tempnam($tempdir, 's3logs_');
+        $s3client->download_file($tempfilepath, $s3logkey);
+        $loghandle = fopen($tempfilepath, 'r');
+        return $loghandle;
+    }
+
+    /**
+     * Gets oldest -> newest array of S3 log keys
+     *
+     * @return array
+     */
+    private static function get_s3_log_keys() {
+        $s3client = new s3_client();
+        $porentialkeys = $s3client->get_all_keys();
+
+        $keys = array();
+        foreach ($porentialkeys as $potentialkey) {
+            // Dont include keys that arn't expected format.
+            if (!preg_match('/.*_\d{14}_\d*_\d*.csv/', $potentialkey)) {
+                continue;
+            }
+
+            // Strip prefix so we can sort.
+            $sortable = trim(substr($potentialkey, strpos($potentialkey, '_') + 1));
+            $keys[$sortable] = $potentialkey;
+        }
+
+        // Sort so ordered oldest -> newest.
+        ksort($keys);
+
+        return $keys;
+    }
+
+    /**
+     * Writes current log contents for courses that are in the DB to log files.
+     *
+     * @param integer $courseids      Course ids.
+     * @param array   $logfilehandles Log file handles, indexed by courseid.
+     * @return void
+     */
+    private static function write_current_database_logs_to_course_logs($courseids, $logfilehandles) {
+        global $DB;
+
+        foreach ($courseids as $courseid) {
+            $records = $DB->get_records(
```

Comment:
Be very careful calling get_records from the standard log store. LTU for example has about a billion rows and is 180GB on disk. 
Really should be using get_recordset and grabbing results by pages. In Moodle 3.5 (and maybe 3,4) you won't need to page with get_recordset see MDL-60174

Reply:
Do you think its worth it given the scale we're dealing with here - only pulling one courses logs at once ?